File: schemas/podcast-platform/convex/reviews.ts
// reviews: User ratings and reviews for shows.
// See README.md for full design rationale.
import { defineTable } from "convex/server";
import { v } from "convex/values";

export const reviews = defineTable({
  userId: v.id("users"),
  showId: v.id("shows"),
  rating: v.number(),
  title: v.optional(v.string()),
  body: v.optional(v.string()),
  updatedAt: v.number(),
})
  .index("by_user_id_show_id", ["userId", "showId"])
  .index("by_show_id_creation_time", ["showId", "_creationTime"])
  .index("by_show_id_rating", ["showId", "rating"]);

File: schemas/podcast-platform/spacetimedb/reviews.rs
// reviews: User ratings and written reviews for shows.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

#[spacetimedb::table(name = reviews, public)]
pub struct Review {
    #[primary_key]
    pub id: String, // UUID
    #[index(btree)]
    pub user_id: String, // UUID — FK → users.id (cascade delete)
    #[index(btree)]
    pub show_id: String, // UUID — FK → shows.id (cascade delete)
    pub rating: i32,
    pub title: Option<String>,
    pub body: Option<String>,
    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}
// Composite unique: (user_id, show_id) — enforce in reducer logic
// Composite index: (show_id, created_at) — not supported, enforce in reducer logic
// Composite index: (show_id, rating) — not supported, enforce in reducer logic
